pkg/kubelet/nodelease/controller.go

```diff
@@ -0,0 +1,176 @@
+	// when renewing the lease before it waits for the renewal interval before trying again,
+	// similar to what we do for node status retries
+	maxUpdateRetries = 5
+)
+
+// Controller manages creating and renewing the lease for this Kubelet
+type Controller interface {
+	Run(stopCh <-chan struct{})
+}
+
+var _ Controller = &controller{}
+
+type controller struct {
+	client                     coordclientset.LeaseInterface
+	holderIdentity             string
+	leaseDurationSeconds       int32
+	renewInterval              time.Duration
+	clock                      clock.Clock
+	onRepeatedHeartbeatFailure func()
+}
+
+// NewController constructs and returns a controller
+func NewController(clock clock.Clock, client clientset.Interface, holderIdentity string, leaseDurationSeconds int32, onRepeatedHeartbeatFailure func()) Controller {
+	var leaseClient coordclientset.LeaseInterface
+	if client != nil {
+		leaseClient = client.CoordinationV1beta1().Leases(metav1.NamespaceNodeLease)
+	}
+	return &controller{
+		client:               leaseClient,
+		holderIdentity:       holderIdentity,
+		leaseDurationSeconds: leaseDurationSeconds,
+		renewInterval:        time.Duration(float64(time.Duration(leaseDurationSeconds)*time.Second) * renewFactor),
+		clock:                clock,
+		onRepeatedHeartbeatFailure: onRepeatedHeartbeatFailure,
+	}
+}
+
+// Run runs the controller
+func (c *controller) Run(stopCh <-chan struct{}) {
+	if c.client == nil {
+		glog.Infof("node lease controller has nil client, will not claim or renew leases")
+		return
+	}
+	wait.Until(c.sync, c.renewInterval, stopCh)
+}
+
+func (c *controller) sync() {
+	lease, created := c.backoffEnsureLease()
+	// we don't need to update the lease if we just created it
+	if !created {
+		c.retryUpdateLease(lease)
+	}
+}
+
+// backoffEnsureLease attempts to create the lease if it does not exist,
+// and uses exponentially increasing waits to prevent overloading the API server
+// with retries. Returns the lease, and true if this call created the lease,
+// false otherwise.
+func (c *controller) backoffEnsureLease() (*coordv1beta1.Lease, bool) {
+	var (
+		lease   *coordv1beta1.Lease
+		created bool
+		err     error
+	)
+	sleep := 100 * time.Millisecond
+	for {
+		lease, created, err = c.ensureLease()
+		if err == nil {
+			break
+		}
+		// log error
+		glog.Errorf("failed to ensure node lease exists, error: %v", err)
+		// backoff wait
+		time.Sleep(sleep)
+		sleep = minDuration(2*sleep, 7*time.Second)
+	}
+	return lease, created
+}
+
+// ensureLease creates the lease if it does not exist. Returns the lease and
+// a bool (true if this call created the lease), or any error that occurs.
+func (c *controller) ensureLease() (*coordv1beta1.Lease, bool, error) {
+	lease, err := c.client.Get(c.holderIdentity, metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		// lease does not exist, create it
+		lease, err := c.client.Create(c.newLease(nil))
```

Comment:
is this working today in a test cluster with authz enabled? I expected to see modifications to the Node authorizer and NodeRestriction admission plugin to let kubelets mess with their own leases

Reply:
Good point, I forgot about that. I will add those changes to this PR, and it probably makes sense to ensure the e2e test also runs in the standard e2e suite, instead of just the e2e node suite.

We also don't currently run the node authorizer/node restriction in the e2e node tests. Maybe I should revisit https://github.com/kubernetes/kubernetes/pull/60172?